pkg/nats/stream/producer_client.go

```diff
@@ -0,0 +1,223 @@
+package stream
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/nats-io/nats.go"
+	"github.com/rs/zerolog/log"
+)
+
+type ProducerClientParams struct {
+	Conn   *nats.Conn
+	Config StreamProducerClientConfig
+}
+
+type ProducerClient struct {
+	Conn *nats.Conn
+	mu   sync.RWMutex // Protects access to activeStreamInfo and activeConnHeartBeatRequestSubjects
+
+	// A map of ConsumerID to StreamId that are active
+	activeStreamInfo map[string]map[string]StreamInfo
+	// A map of ConsumerID to the subject where a heartBeatRequest needs to be sent.
+	activeConnHeartBeatRequestSubjects map[string]string
+	heartBeatCancelFunc                context.CancelFunc
+	config                             StreamProducerClientConfig
+}
+
+func NewProducerClient(ctx context.Context, params ProducerClientParams) (*ProducerClient, error) {
+	nc := &ProducerClient{
+		Conn:                               params.Conn,
+		activeStreamInfo:                   make(map[string]map[string]StreamInfo),
+		activeConnHeartBeatRequestSubjects: make(map[string]string),
+		config:                             params.Config,
+	}
+
+	go nc.heartBeat(ctx)
+
+	return nc, nil
+}
+
+func (pc *ProducerClient) AddStream(
+	consumerID string,
+	streamID string,
+	requestSub string,
+	heartBeatRequestSub string,
+) error {
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+
+	streamInfo := StreamInfo{
+		ID:         streamID,
+		RequestSub: requestSub,
+		CreatedAt:  time.Now(),
+	}
+
+	if pc.activeStreamInfo[consumerID] == nil {
+		pc.activeStreamInfo[consumerID] = make(map[string]StreamInfo)
+	}
+
+	if _, ok := pc.activeStreamInfo[consumerID][streamID]; ok {
+		return fmt.Errorf("cannot create request with same streamId %s again", streamID)
+	}
+
+	pc.activeStreamInfo[consumerID][streamID] = streamInfo
+	pc.activeConnHeartBeatRequestSubjects[consumerID] = heartBeatRequestSub
+
+	return nil
+}
+
+func (pc *ProducerClient) RemoveStream(consumerID string, streamID string) {
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+
+	activeStreamIdsForConn, ok := pc.activeStreamInfo[consumerID]
+	if !ok {
+		return
+	}
+
+	if _, ok := activeStreamIdsForConn[streamID]; !ok {
+		return
+	}
+
+	delete(activeStreamIdsForConn, streamID)
+
+	if len(activeStreamIdsForConn) == 0 {
+		delete(pc.activeStreamInfo, consumerID)
+		delete(pc.activeConnHeartBeatRequestSubjects, consumerID)
+	}
+}
+
+func (pc *ProducerClient) heartBeat(ctx context.Context) {
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	pc.heartBeatCancelFunc = cancel
+
+	ticker := time.NewTicker(pc.config.HeartBeatIntervalDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctxWithCancel.Done():
+			log.Ctx(ctxWithCancel).Debug().Msg("Heart beat for NATS based streaming producer client cancelled.")
+			return
+		case <-ticker.C:
+
+			nonActiveStreamIds := make(map[string][]string)
+
+			for c, v := range pc.activeConnHeartBeatRequestSubjects {
```

Comment:
we need to be careful about excessive locking, which the code as it stands already does as it is locking for each message that needs to be sent for _any_ stream managed by the producer. The plan is for Udit to create a followup issue to improve the concurrency and locking, but can't find the issue at the moment.

Overall, we don't care about consistency of streams while heartbeating as the consumer is adding a graceful window, but not sure about how golang maps handle concurrency access. If they panic, then thats a problem. 

The desired state we want to get is:
- Write lock when adding or removing streams
- Read lock during heartbeats
- Don't lock when writing messages 

The last one is tricky as it requires a decoupling stream writers from producers. The producer client should return writer instance to the ComputeHandler, and then we write directly to it without any locking. When the stream is deleted, then the producer client will close the writer and all future writes will be rejected.

Reply:
On the third point. Having the write instance in `ComputeHandler` makes things difficult. You ideally do not want to write to the subject if the stream is no longer active and if that information is obtained from the heart beat. However, that information is available, only to the producer client and hence makes it difficult to decouple the writer instance from the client and provide it to the handler.

Reply:
@wdbaruni  It does panic on on concurrent writes. To test your hypothesis I wrote this small function 

`func main() {
    m := make(map[int]int)
    go writeToMap(m)
    go writeToMap(m)

    time.Sleep(time.Second)
}

func writeToMap(m map[int]int) {
    for i := 0; i < 1000; i++ {
        m[i] = i
    }
}`

In addition, it also panics in case a go routine is writing to a map and then some go routine tries to read during this time. Though this happens sporadically.


`package main

import (
	"fmt"
	"time"
)

func main() {
	m := make(map[int]int)

	go func() {
		for i := 0; i < 1000; i++ {
			m[i] = i
		}
	}()

	go func() {
		for i := 0; i < 1000; i++ {
			fmt.Println(m[i])
		}
	}()

	time.Sleep(time.Second)
}`

Reply:
The one way to achieve this, 

> Don't lock when writing messages

would be to have 2 channels in compute handler. 

1. Where you receive logs from engines, like docker etc
2. From Producer client which streams inactive stream id messages to compute handler.
   
We do a select between the 2 channels. And if the inactive stream id is receieved in compute handler from producer client we don't write message. This way you are not holding the lock while writing messages.

Please note this is very high level thinking, I ll dig bit more to see if this is actually achievable. 

Let me know if you see some obvious flaw with it, which i may be missing here.